src/solver/cartesianprod/backward_cartesianprod.cpp

```diff
@@ -0,0 +1,167 @@
+}
+
+bool CartesianProdBackward::IsApplicable(
+    const ExecutionContext&, const miopen::cartesianprod::BwdProblemDescription& problem) const
+{
+    if(!(problem.GetOutputGradDesc().GetType() == miopenHalf ||
+         problem.GetOutputGradDesc().GetType() == miopenFloat ||
+         problem.GetOutputGradDesc().GetType() == miopenBFloat16))
+    {
+        return false;
+    }
+    if(!problem.IsAllPacked())
+    {
+        return false;
+    }
+    if(!IsOverRocmBwd(problem))
+    {
+        return false;
+    }
+    return true;
+}
+
+ConvSolution CartesianProdBackward::GetSolution(
+    const ExecutionContext& context,
+    const miopen::cartesianprod::BwdProblemDescription& problem) const
+{
+    std::ignore = context;
+
+    auto dtype        = problem.GetOutputGradDesc().GetType();
+    auto output_dtype = miopen::GetDataType(dtype);
+    auto inputCount   = problem.GetInputCount();
+
+    auto build_params =
+        KernelBuildParameters{{"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+                              {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+                              {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+                              {"TILE_SIZE", TILE_SIZE},
+                              {"D_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype}};
+
+    if(inputCount == 1)
+    {
+        return ConvSolution{miopenStatusNotImplemented};
+    }
+
+    auto result = ConvSolution{miopenStatusSuccess};
+    for(int i = inputCount - 1; i >= 0; i--)
+    {
+        auto input_grad_size = problem.GetInputGradDesc(i).GetElementSize();
+        result.construction_params.push_back(make_hip_kernel({LOCAL_SIZE_BWD},
+                                                             {input_grad_size},
+                                                             "MIOpenCartesianProd.cpp",
+                                                             "CartesianProdBackward",
+                                                             build_params));
+    }
+
+    result.invoker_factory = [inputCount](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            size_t stride         = 1;
+            int kernelCnt         = 0;
+            decltype(auto) params = raw_params.CastTo<miopen::cartesianprod::BwdInvokeParams>();
+            auto output_grad_tv   = get_inner_expanded_tv<2>(deref(params.outputGradDesc));
+            HipEventPtr start, stop;
+            bool profiling = handle_.IsProfilingEnabled();
+            if(profiling)
+            {
+                handle_.EnableProfiling(false);
+                hipStreamSynchronize(handle_.GetStream());
+                start = miopen::make_hip_event();
+                stop  = miopen::make_hip_event();
+                hipEventRecord(start.get(), handle_.GetStream());
+            }
+
+            par_for(inputCount, [&](auto i) {
```

Comment:
Parallelism is meaning less, because all kernels are scheduled on the same stream.

Reply:
resolved